utils/pattern/src/replacement.rs

```diff
@@ -0,0 +1,215 @@
+///     Element::TokenZero
+/// ]);
+/// replacements.insert(5, vec![
+///     Element::TokenFive
+/// ]);
+///
+/// let mut parser = Parser::new("{5}, {0}", false);
+/// let mut interpolator = Interpolator::new(parser, replacements);
+///
+///
+/// assert_eq!(Ok(Some(Element::TokenFive)), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::TokenZero)), interpolator.try_next());
+/// assert_eq!(Ok(None), interpolator.try_next());
+/// ```
+///
+/// # Advanced Replacement Provider
+///
+/// Another scenario in which a consumer may want to implement their own `ReplacementProvider`
+/// is when they can avoid allocation and storing of the elements, and instead can stream the
+/// replacements or compute them on the fly.
+///
+/// To illustrate such an example, we'll create a custom iterator which computes digits from `1`
+/// to some `max` value and return such iterator from a custom replacement provider.
+///
+/// ## Examples
+/// ```
+/// use icu_pattern::{Parser, Interpolator, ReplacementProvider};
+///
+/// #[derive(Debug, PartialEq)]
+/// enum Element {
+///     Digit(usize),
+///     Literal(String)
+/// }
+///
+/// impl From<&str> for Element {
+///     fn from(input: &str) -> Self {
+///         Self::Literal(input.to_string())
+///     }
+/// }
+///
+/// struct MyReplacementProvider;
+///
+/// struct MyIterator {
+///     idx: usize,
+///     max: usize,
+/// }
+///
+/// impl MyIterator {
+///     pub fn new(max: usize) -> Self {
+///         Self {
+///             idx: 0,
+///             max
+///         }
+///     }
+/// }
+///
+/// impl Iterator for MyIterator {
+///     type Item = Element;
+///
+///     fn next(&mut self) -> Option<Self::Item> {
+///         self.idx += 1;
+///         if self.idx <= self.max {
+///             Some(Element::Digit(self.idx))
+///         } else {
+///             None
+///         }
+///     }
+/// }
+///
+///
+/// impl ReplacementProvider<usize, Element> for MyReplacementProvider {
+///     type Iter = MyIterator;
+///
+///     fn take_replacement(&mut self, key: &usize) -> Option<Self::Iter> {
+///         Some(MyIterator { idx: 0, max: *key })
+///     }
+/// }
+///
+/// let mut parser = Parser::new("{4}, {2}", false);
+/// let mut interpolator = Interpolator::new(parser, MyReplacementProvider);
+///
+/// assert_eq!(Ok(Some(Element::Digit(1))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(2))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(3))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(4))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(1))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(2))), interpolator.try_next());
+/// assert_eq!(Ok(None), interpolator.try_next());
+/// ```
+pub trait ReplacementProvider<K, E> {
```

Comment:
I think both of these parameters can be an associated type; and this will _greatly_ simplify the generics on Interpolator.

Reply:
I guess that does not allow you to have both `Vec<E>` and `Vec<Vec<E>>` -- in that case at least the key type should be made an associated type IMO.

Reply:
Done.